marconi-chain-index/src/Marconi/ChainIndex/Indexers/EpochStakepoolSize.hs

```diff
@@ -1,187 +1,502 @@
-  let
-    lastTwo (_, a, b) = (a, b)
-    result = all'
-      & groupBy ((==) `on` (\(e, _, _) -> e))
-      & map (\case xs@((e, _, _) : _) -> Just $ Event (e, M.fromList $ map lastTwo xs); _ -> Nothing)
-      & P.catMaybes
-  pure result
+        SQL.execute_ c "BEGIN"
+        forM_ (concatMap eventToEpochSPDRows $ filter epochSPDEventIsFirstEventOfEpoch eventsList) $ \row ->
+          SQL.execute c
+              [r|INSERT INTO epoch_spd
+                  ( epochNo
+                  , poolId
+                  , lovelace
+                  , slotNo
+                  , blockHeaderHash
+                  , blockNo
+                  ) VALUES (?, ?, ?, ?, ?, ?)|] row
+        SQL.execute_ c "COMMIT"
+
+        -- We store the LedgerState if one of following conditions hold:
+        --   * the LedgerState cannot be rollbacked and is the last of an epoch
+        --   * the LedgerState can be rollbacked
+        let writeLedgerState ledgerState (C.SlotNo slotNo) blockHeaderHash (C.BlockNo blockNo) isRollbackable = do
+                let fname = ledgerStateDirPath
+                        </> "ledgerState_"
+                         <> (if isRollbackable then "volatile_" else "")
+                         <> show slotNo
+                         <> "_"
+                         <> Text.unpack (C.serialiseToRawBytesHexText blockHeaderHash)
+                         <> "_"
+                         <> show blockNo
+                         <> ".bin"
+                -- TODO We should delete the file is the write operation was interrumpted by the
+                -- user. Tried using something like `onException`, but it doesn't run the cleanup
+                -- function. Not sure how to do the cleanup here without restoring doing it outside
+                -- the thread where this indexer is running.
+                BS.writeFile fname (CBOR.toLazyByteString $ encodeLedgerState ledgerState)
+        forM_ eventsList
+              $ \(EpochSPDEvent
+                    maybeLedgerState
+                    maybeEpochNo
+                    _
+                    slotNo
+                    blockHeaderHash
+                    blockNo
+                    chainTip
+                    isFirstEventOfEpoch) -> do
+            case (maybeEpochNo, maybeLedgerState) of
+              (Just _, Just ledgerState) -> do
+                  let isRollbackable = isBlockRollbackable securityParam blockNo chainTip
+                  when (isRollbackable || isFirstEventOfEpoch) $ do
+                    writeLedgerState ledgerState slotNo blockHeaderHash blockNo isRollbackable
+              -- We don't store any 'LedgerState' if the era doesn't have epochs (Byron era) or if
+              -- we don't have access to the 'LedgerState'.
+              _noLedgerStateOrEpochNo -> pure ()
+
+        -- Remove all immutable LedgerStates from the filesystem expect the most recent immutable
+        -- one which is from the last slot of latest epoch.
```

Comment:
Now it's the first slot of the next epoch (right?).

Reply:
289542b0de9313055702bbafa38624783c00032d